File: guests/go/menus/menus.go
// The menus conformance scene (tools/scenes/menus.steps). Canonical
// semantics in guests/rust/menus.rs.
package menus

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

func App() *kaya.App {
	app := kaya.NewApp()

	var (
		groups kaya.Collection
		items  kaya.Collection
	)

	app.Build(func(tx *kaya.Tx) {
		status := tx.Signal("ready")
		canExport := tx.Signal(false)
		details := tx.Signal(false)
		sort := tx.Signal(0.0)

		onShare := func(tx *kaya.Tx) { tx.Write(status, "shared") }

		win := tx.Window(0).Title("menus")
		file := win.Menu("File").BindEnabled(canExport)
		// The vocabulary has no save glyph, so `done` is the spelling.
		file.Item("Save").Symbol(kaya.SymbolDone).Shortcut("primary+s").
			OnActivate(func(tx *kaya.Tx) {
				tx.Write(status, "saved")
			})
		file.Item("Export").BindEnabled(canExport).Symbol(kaya.SymbolForward)
		share := file.Item("Share").Primary(true).OnActivate(onShare)

		win.Menu("View").Toggle("Details").BindChecked(details).
			Symbol(kaya.SymbolInfo).
			OnToggle(func(tx *kaya.Tx, on bool) {
				if on {
					tx.Write(status, "details on")
				} else {
					tx.Write(status, "details off")
				}
			})

		// Option order IS the index.
		sortGroup := win.RadioGroup("Sort")
		sortGroup.Option("Name")
		sortGroup.Option("Date")
		sortGroup.BindValue(sort).OnSelect(func(tx *kaya.Tx, index int) {
			if index == 1 {
				tx.Write(status, "sorted date")
			} else {
				tx.Write(status, "sorted name")
			}
		})

		groups = tx.Collection()
		catalog := tx.ContextCatalog()
		catalog.Item("Remove").Symbol(kaya.SymbolDelete).
			OnActivateNode(func(tx *kaya.Tx, keys []any) {
				group, item := keys[0].(string), keys[1].(string)
				tx.Remove(items.At(group), item)
				tx.Write(status, fmt.Sprintf("removed %s/%s", group, item))
			})

		tx.Mount(tx.Column(func() {
			tx.Label(status) // label#0
			tx.Button("enable export", func(tx *kaya.Tx) { // button#0
				tx.Write(canExport, true)
			})
			tx.Button("reset menu state", func(tx *kaya.Tx) { // button#1
				// The folds never echo, so these reset the user-state mirror.
				tx.Write(details, false)
				tx.Write(sort, 0.0)
				tx.Write(status, "ready")
			})
			tx.Button("extend menus", func(tx *kaya.Tx) { // button#2
				tx.Menu(share).Primary(false)
				tx.Menu(file).Label("Document").
					Item("Publish").Primary(true).Symbol(kaya.SymbolCopy).
					OnActivate(onShare)
				tx.Window(0).Menu("Tools").Item("Inspect").Symbol(kaya.SymbolSearch)
			})

			targetText := tx.Signal("rename target")
			target := tx.Label(targetText) // label#1
			tx.ContextMenu(target).Item("Rename").Symbol(kaya.SymbolEdit).
				OnActivate(func(tx *kaya.Tx) {
					tx.Write(status, "renamed")
				})

			for g := range tx.Rows(groups).All() {
				g.Column(func() {
					items = g.Collection()
					for row := range g.Rows(items).All() {
						// label#2 once g2/a stamps
						row.ContextMenu(row.Label(row.Value()), catalog)
					}
				})
			}
		}))
	})

	// Seed after mount: the stamp path attaches the shared catalog and keys.
	app.Build(func(tx *kaya.Tx) {
		tx.Insert(groups, "g2", "Home")
		tx.Insert(items.At("g2"), "a", "water plants")
	})

	return app
}

File: guests/java/dev/kaya/guests/Menus.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

/**
 * The menus conformance scene, JVM port: the command vocabulary (a
 * File/View/Sort menu bar, context menus on a live label and on stamped
 * rows), the uncontrolled-menu echo doctrine, and a late
 * rename/append/promotion rework. Canonical semantics in
 * guests/rust/menus.rs; the byte-frozen contract in tools/scenes/menus.steps.
 */
public final class Menus {
    // Java lambdas cannot assign captured locals; these live here for the
    // seed and the Remove fold.
    private static KayaApp.Collection groups;
    private static KayaApp.Collection items;

    public static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            KayaApp.Signal<String> status = tx.signal("ready");
            KayaApp.Signal<Boolean> canExport = tx.signal(false);
            KayaApp.Signal<Boolean> details = tx.signal(false);
            KayaApp.Signal<Double> sort = tx.signal(0.0);

            java.util.function.Consumer<KayaApp.Tx> onShare =
                    t -> t.write(status, "shared");

            KayaApp.WindowRef win = tx.window(0).title("menus");
            KayaApp.MenuItem file = win.menu("File").enabled(canExport);
            // Symbols are CONCEPTS, drawn per platform: the vocabulary
            // has no `save`, so `DONE` is the checkmark idiom
            // (docs/styling-plan.md D6).
            file.item("Save")
                    .symbol(KayaApp.Symbol.DONE)
                    .shortcut("primary+s")
                    .onActivate(t -> t.write(status, "saved"));
            file.item("Export").enabled(canExport).symbol(KayaApp.Symbol.FORWARD);
            KayaApp.MenuItem share = file.item("Share").primary(true).onActivate(onShare);

            win.menu("View").toggle("Details").checked(details)
                    .symbol(KayaApp.Symbol.INFO)
                    .onToggle((t, on) ->
                            t.write(status, on ? "details on" : "details off"));

            // Option order IS the index vocabulary: Name = 0, Date = 1.
            KayaApp.MenuItem sortGroup = win.radioGroup("Sort");
            sortGroup.option("Name");
            sortGroup.option("Date");
            sortGroup.value(sort).onSelect((t, index) ->
                    t.write(status, index == 1 ? "sorted date" : "sorted name"));

            groups = tx.collection();
            KayaApp.ContextCatalog catalog = tx.contextCatalog();
            catalog.item("Remove").symbol(KayaApp.Symbol.DELETE)
                    .onActivateNode((t, keys) -> {
                        String group = (String) keys.get(0);
                        String item = (String) keys.get(1);
                        t.remove(items.at(group), item);
                        t.write(status, "removed " + group + "/" + item);
                    });

            tx.mount(tx.column(() -> {
                tx.label(status); // label#0
                tx.button("enable export", t -> // button#0
                        t.write(canExport, true));
                tx.button("reset menu state", t -> { // button#1
                                        // The folds never echo the user's pick, so these
                                        // writes are real records (never coalesced) that
                                        // reset the backend's user-state mirror.
                    t.write(details, false);
                    t.write(sort, 0.0);
                    t.write(status, "ready");
                });
                tx.button("extend menus", t -> { // button#2
                    // Append-only: rename the retained File, move the
                    // promotion hint, grow the bar by Tools.
                    t.menu(share).primary(false);
                    t.menu(file).label("Document")
                            .item("Publish").primary(true)
                            .symbol(KayaApp.Symbol.COPY)
                            .onActivate(onShare);
                    t.window(0).menu("Tools").item("Inspect")
                            .symbol(KayaApp.Symbol.SEARCH);
                });

                KayaApp.Signal<String> targetText = tx.signal("rename target");
                KayaApp.Widget target = tx.label(targetText); // label#1
                tx.contextMenu(target).item("Rename")
                        .symbol(KayaApp.Symbol.EDIT)
                        .onActivate(t -> t.write(status, "renamed"));

                for (var g : tx.rows(groups)) {
                    g.column(() -> {
                        items = g.collection();
                        for (var row : g.rows(items)) {
                            // label#2 once g2/a stamps
                            row.contextMenu(row.label(row.value()), catalog);
                        }
                    });
                }
            }));
            return null;
        });

        // Seed after mount: the stamp path attaches the shared catalog and keys.
        app.build(tx -> {
            tx.insert(groups, "g2", "Home");
            tx.insert(items.at("g2"), "a", "water plants");
            return null;
        });

        app.dispatchLoop();
    }

    private Menus() {}
}
